src/persist-txn/src/txn_write.rs

```diff
@@ -376,8 +376,11 @@ mod tests {
         log.record_txn(2, &txn);
         assert_eq!(apply_2.is_empty(), false);
         cache.update_gt(&2).await;
+        // Manually delete the register from unapplied registers since there's no event to signal
```

Comment:
Instead of this, could we do something like:

- register at 1 (as above)
- write something empty at 2
- compact the TxnsHandle to 2 (which should go through and physically advance the since)
- _then_ start up the TxnsCache at 2 (instead of 0), which would get the register at ts=1 but compacted_ts=2 and ignore it

An alternative to that that's maybe a touch more direct is to plus the ts given to `expect_open` all the way to the Snapshot's as_of. That would let you skip the "compact the TxnsHandle to 2" step because we could just expect_open at 2 and it would get data with advanced timestamps

This is also totally fine as you have it

Reply:
Interestingly the following

```Rust
        let mut txns = TxnsHandle::expect_open(PersistClient::new_for_tests().await).await;
        let log = txns.new_log();
        let d0 = txns.expect_register(1).await;
        txns.expect_commit_at(2, d0, &[], &log).await;
        txns.compact_to(2).await;
        let mut cache = TxnsCache::expect_open(2, &txns).await;
        assert_eq!(cache.min_unapplied_ts(), &3);
```
fails with the following error:

```Rust
assertion `left == right` failed
  left: 1
 right: 3
```

When opening the cache, the register event still has a compacted timestamp of `1` so it's added to the unapplied registers. Am I missing something obvious or is something weird going on here?

(Note: I know the `txns.compact_to(2).await;` is unnecessary because it's included in the `expect_open`, I just did it as a sanity check).

Reply:
Ah, doh, it's because the TxnsHandle has a Subscribe, which has its own since capability. I think the "alternative" where we plumb the ts given to `expect_open` down to be used as the as_of for the Snapshot is more foolproof

Reply:
How would we do that? It looks like we'd have to change the `TxnsCache::open` method which feels pretty invasive.

(Happy to huddle about this if PR comments are too slow to iterate.)

Reply:
Yeah, we'd have to change `open`. I think we should stick with what you originally had and get this merged. I'd love to rebase this in to my ci branch and then kick off another nightlies to run while I'm skunkworks-ing today